### ISSUE TITLE
[ENH] `deepDNAshape` reworked

### ISSUE THREAD

**Issue by fkiraly** [gc-os-ai/pyaptamer#110]
For DeepAptamer, one of the required components is the feature extractor (deep)DNAshape.

As per discussion in https://github.com/gc-os-ai/pyaptamer/issues/82#issuecomment-3193622643 and following, a package exists
https://github.com/JinsenLi/deepDNAshape

however it is not directly usable for us:

* it is not a pypi installable package
* it is `tensorflow`, not `torch` based
* it requires a comparably ancient `numpy<1.24`

In order to enable the reimplementation of DeepAptamer, we have hence discussed that the only feasible way forward (with an algorithm that includes deepDNAshape) is forking the deepDNAshape package into our library.

In the process of this, translating to `torch` to remove the version conflicts.

Crediting and license is straightforward:

* license is BSD-3, permissive. We need to credit code that is copied or inspired by
* in docstrings as well, we whould mention the original author, @jinsenli to give clear credit, and cite the original publication

**Comment by satvshr** [comment]
@fkiraly I also think we can involve [Rohs lab](https://www.rohslab.org/)(if they are interested) in this if they are interested, given they are huge (and maybe only) contributors in the whole "DNA shape" space.

**Comment by fkiraly** [comment]
I see - they are distinct from the deepaptamer authors, right?

**Comment by avinab-neogy** [comment]
Hi @satvshr @fkiraly  ,
I am working on this issue now. 
I went through #82 and have an understanding of how the `deepDNAshape` does not use the pentamer sliding window to calculate the MGW, Propeller twist, Helical twist and Roll so it returns the entire 138 features unlike the `DNAshape` library.

I also went through #98 and it is parked till this PR is merged right so should i go through that code?
thanks for the suggestion of `onnx` seems quite useful.

Will be updating my progress here.

**Comment by avinab-neogy** [comment]
> does it not? I thought the layer argument is the sliding window?

You are right actually it still uses a sliding window but a variable one

`deepDNAshape --seq AGAGATACGATACGA --feature ProT --layer 7`

This uses a 2bp flanking region on each side, resulting in a 15-base receptive field (7*2 + 1). ( the `DNAshape uses a fixed 5-base receptive field)
At positions near the sequence ends (e.g., i=1), some flanking positions are missing. However, thanks to the deep neural network’s convolutional filters and padding, `deepDNAshape` can still generate predictions by learning from partial context.
Increasing the --layer value expands the receptive field, providing more context ( on the other end ) especially helpful near edges, allowing reliable feature calculations even when flanking bases are limited.

Thanks Satvik, I understand this method better now! 

> As mentioned above, I would suggest looking at the predictor call only given that is all that we need.

Thanks a lot for specifying!

**Comment by aditi-dsi** [comment]
Hi @satvshr, @fkiraly, since this issue is inactive, I am picking it up to work on (hope it's not invalid now?). Will update the progress here.

**Comment by prashantpandeygit** [comment]
hi @aditi-dsi  just letting you know that i'm working on this too, i am halfway through; maybe you could review my PR, if it is not sufficient you may pull up a PR then? would that be fine?? else it would be unnecessary extra work, just bear with me for a day or two, my bad for not mentioning it earlier here. hope you understand!

**Comment by aditi-dsi** [comment]
@prashantpandeygit sure, go ahead. Since your work is in progress, I'd suggest opening a draft PR to avoid duplicate work.
